### ISSUE TITLE
mypy doesn't work nicely with lambdas returning named tuples 

### ISSUE THREAD

**Issue by david-gang** [python/mypy#5069]
Note: if you are reporting a wrong signature of a function or a class in
the standard library, then the typeshed tracker is better suited
for this report: https://github.com/python/typeshed/issues

Please provide more information to help us understand the issue:

* Are you reporting a bug, or opening a feature request? bug
* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.
```
from typing import NamedTuple


class A(NamedTuple):
    '''
    value class combining origin coordinate with list of destination coordinates
    '''
    a: str


def foo() -> None:
    A("david")
    greet = lambda a: A("hi " + a)
    greet('moshe')
```
* What is the actual behavior/output?
a.py:14: error: Call to untyped function (unknown) in typed context
* What is the behavior/output you expect?
mypy should actually be smart enough to see that i am returning a named tuple: http://mypy.readthedocs.io/en/latest/kinds_of_types.html#callable-types-and-lambdas
> Lambdas are also supported. The lambda argument and return value types cannot be given explicitly; they are always inferred based on context using bidirectional type inference:

* What are the versions of mypy and Python you are using?
the latest and greatest: mypy 0.600 and python 3.6.5
  Do you see the same issue after installing mypy from Git master? Didn't try
* What are the mypy flags you are using? (For example --strict-optional)
--strict

**Comment by JelleZijlstra** [comment]
I don't think this is related to namedtuples. Here's a simpler example:
```python
def g(x: str) -> str:
    return x

def foo() -> None:
    greet2 = lambda a: g(a)
    reveal_type(greet2)  # E: Revealed type is 'def (a: Any) -> builtins.str'
    greet2("jelle")  # E: Call to untyped function (unknown) in typed context
```
There are two issues here:
- We don't infer the argument type of the lambda, even though it is unambiguous.
- The error message is not very clear; it should probably say "<lambda>" instead of "unknown".

**Comment by DetachHead** [comment]
the issue doesn't seem to have anything to do with the lambda's arguments being `Any`. here's another example

```py
from typing import Callable
(lambda: 1)()  # no error
a = lambda: 1
b: Callable[[], int] = lambda: 1
reveal_type(a)  # Revealed type is "def () -> builtins.int"
reveal_type(b)  # Revealed type is "def () -> builtins.int"
a()  # error: Call to untyped function (unknown) in typed context  [no-untyped-call]
b()  # no error
```

**Comment by steve-mavens** [comment]
For @DetachHead 's example, it seems too strict that the type `"def () -> builtins.int"` is considered to be a call to an untyped function, when in fact it's fully typed since there are no input parameters to worry about.

**Comment by DetachHead** [comment]
@steve-mavens raised #14473
